binaryen-sys/build.rs

```diff
@@ -23,13 +30,122 @@ fn gen_bindings() {
+    let re = Regex::new(r#"registerPass\("([^"]+)", "([^"]+)", [^)]+\);"#).unwrap();
+
+    let mut passes: Vec<Pass> = vec![];
+
+    let input = File::open("binaryen/src/passes/pass.cpp").expect("Couldn't open pass.cpp");
+    for line in BufReader::new(input).lines() {
+        let line = line.unwrap();
+        let caps = re.captures(&line);
+        if caps.is_some() {
+            let caps = caps.unwrap();
+            let name = caps.get(1).unwrap().as_str();
+            let description = caps.get(2).unwrap().as_str();
+
+            passes.push(Pass {
+                id: name.to_camel_case(),
+                name: name.to_string(),
+                description: description.to_string(),
+            });
+        }
+    }
+
+    passes
+}
+
+fn gen_passes() {
+    let passes: Vec<Pass> = read_passes();
+
+    let ids: Vec<String> = passes
+        .iter()
+        .map(|pass| pass.id.to_string())
+        .collect();
+
+    let fromstrs: Vec<String> = passes
+        .iter()
+        .map(|pass| format!(r#""{}" => Ok(OptimizationPass::{})"#, pass.name.to_string(), pass.id.to_string()))
+        .collect();
+
+    let descriptions: Vec<String> = passes
+        .iter()
+        .map(|pass| format!(r#"OptimizationPass::{} => "{}""#, pass.id.to_string(), pass.description.to_string()))
+        .collect();
+
+    let output = format!(r#"
+        use std::str::FromStr;
+
+        #[derive(PartialEq, Debug)]
+        enum OptimizationPass {{
+            {ids}
+        }}
+
+        impl FromStr for OptimizationPass {{
+            type Err = ();
+            fn from_str(s: &str) -> Result<Self, Self::Err> {{
+                match s {{
+                    {fromstrs},
+                    _ => Err(()),
+                }}
+            }}
+        }}
+
+        trait OptimizationPassDescription {{
+            fn description(&self) -> &'static str;
+        }}
+
+        impl OptimizationPassDescription for OptimizationPass {{
+            fn description(&self) -> &'static str {{
+                match self {{
+                    {descriptions}
+                }}
+            }}
+        }}
+
+        #[cfg(test)]
+        mod tests {{
+            use super::*;
+
+            #[test]
+            fn test_from_str() {{
+                assert_eq!(OptimizationPass::{test_id}, OptimizationPass::from_str("{test_name}").expect("from_str expected to work"));
+            }}
+
+            #[test]
+            fn test_description() {{
+                assert_eq!(OptimizationPass::{test_id}.description(), "{test_description}");
+            }}
+        }}
+    "#,
+    ids = ids.join(", "),
+    fromstrs = fromstrs.join(", "),
+    descriptions = descriptions.join(", "),
+    test_id = passes[0].id.to_string(),
+    test_name = passes[0].name.to_string(),
+    test_description = passes[0].description.to_string()
+    );
+
+    println!("{}", output);
+
+    fs::write("src/passes.rs", output).expect("Unable to write passes.rs");
```

Comment:
In general build scripts should try to avoid modifying any files except from `OUT_DIR`. It is especially important to not alter any files in the source directory, since cargo expects the crate directories to be immutable. 

I think running `cargo package` should tell you about the problem.

Reply:
Let me know where to generate the source to.

Reply:
You can use `env::var("OUT_DIR")` to get the directory that is devoted for auto-generated code. Then, you can add the generated code via include, like we are doing now for bindings generated by `bindgen`.

https://github.com/pepyakin/binaryen-rs/blob/28fcad4feb6fe5e57fde7198fcfd1cc3aa3c5df9/binaryen-sys/src/lib.rs#L5

Reply:
Changed it, please check.